src/views/Nft/market/hooks/useCollectionNfts.tsx

```diff
@@ -0,0 +1,268 @@
+  const attrParams = Object.values(filterObject).reduce(
+    (accum, attr) => ({
+      ...accum,
+      [attr.traitType]: attr.value,
+    }),
+    {},
+  )
+  const attrFilters = !isEmpty(attrParams) ? await fetchNftsFiltered(address, attrParams) : null
+  return attrFilters ? Object.values(attrFilters.data).map((apiToken) => apiToken.tokenId) : null
+}
+
+const fetchMarketDataNfts = async (
+  collection: Collection,
+  settings: ItemListingSettings,
+  page: number,
+  tokenIdsFromFilter: string[],
+): Promise<NftToken[]> => {
+  const whereClause = tokenIdsFromFilter
+    ? {
+        collection: collection.address.toLowerCase(),
+        isTradable: true,
+        tokenId_in: tokenIdsFromFilter,
+      }
+    : { collection: collection.address.toLowerCase(), isTradable: true }
+  const subgraphRes = await getNftsMarketData(
+    whereClause,
+    REQUEST_SIZE,
+    settings.field,
+    settings.direction,
+    page * REQUEST_SIZE,
+  )
+
+  const apiRequestPromises = subgraphRes.map((marketNft) => getNftApi(collection.address, marketNft.tokenId))
+  const apiResponses = await Promise.all(apiRequestPromises)
+  const newNfts: NftToken[] = apiResponses.reduce((acc, apiNft) => {
+    if (apiNft) {
+      acc.push({
+        ...apiNft,
+        collectionAddress: collection.address,
+        collectionName: apiNft.collection.name,
+        marketData: subgraphRes.find((marketNft) => marketNft.tokenId === apiNft.tokenId),
+      })
+    }
+    return acc
+  }, [] as NftToken[])
+  return newNfts
+}
+
+const tokenIdsFromFallback = (
+  collection: Collection,
+  tokenIdsFromFilter: string[],
+  fetchedNfts: NftToken[],
+  fallbackPage: number,
+): string[] => {
+  let tokenIds: string[] = []
+  const startIndex = fallbackPage * REQUEST_SIZE
+  const endIndex = (fallbackPage + 1) * REQUEST_SIZE
+  if (tokenIdsFromFilter) {
+    tokenIds = tokenIdsFromFilter
+      .filter((tokenId) => !fetchedNfts.some((fetchedNft) => fetchedNft.tokenId === tokenId))
+      .slice(startIndex, endIndex)
+  } else {
+    const totalSupply = parseInt(collection.totalSupply)
+    let counter = startIndex
+    let index = startIndex
+    while (counter < endIndex) {
+      if (index > totalSupply) {
+        break
+      }
+      // eslint-disable-next-line no-loop-func
+      if (!fetchedNfts.some((fetchedNft) => parseInt(fetchedNft.tokenId) === index)) {
+        tokenIds.push(index.toString())
+        counter++
+      }
+      index++
+    }
+  }
+  return tokenIds
+}
+
+const fetchAllNfts = async (
+  collection: Collection,
+  settings: ItemListingSettings,
+  page: number,
+  tokenIdsFromFilter: string[],
+  fetchedNfts: NftToken[],
+  fallbackMode: React.MutableRefObject<boolean>,
+  fallbackPage: React.MutableRefObject<number>,
+): Promise<NftToken[]> => {
+  const newNfts: NftToken[] = []
+  let tokenIds: string[] = []
+  let collectionNftsResponse: ApiResponseCollectionTokens = null
+  if (settings.field !== 'tokenId' && !fallbackMode.current) {
```

Comment:
hmm. Allowing `fetchAllNfts` to know about `useRef` is quite strange to me. The function handles more responsibilities than it should be, I think.

Reply:
Done